tests/baselines/reference/emit(jsx=preserve).js

```diff
@@ -0,0 +1,131 @@
+import {} from "foo.ts";
+import {} from "pkg/foo.ts";
+import {} from ".foo.ts";
+import {} from "./foo.d.ts";
+import {} from "./foo.d.mts";
+import {} from "./foo.d.css.ts";
+import {} from "#internal/foo.ts";
+import {} from "node:foo.ts";
+(require)("./foo.ts");
+import("node:path");
+require("node:path");
+
+//// [lol.ts]
+// Sad face https://github.com/microsoft/TypeScript/blob/6b04f5039429b9d412696fe2febe39ecc69ad365/src/testRunner/compilerRunner.ts#L207
+
+
+//// [main.js]
+var __rewriteRelativeImportExtension = (this && this.__rewriteRelativeImportExtension) || function (path, preserveJsx) {
+    if (typeof path === "string" && path[0] === "." && (path[1] === "/" || path[1] === "." && path[2] === "/")) {
+        if (path.substring(path.length - 4) === ".tsx") {
+            return path.substring(0, path.length - 4) + (preserveJsx ? ".jsx" : ".js");
+        }
+        if (path.substring(path.length - 3) === ".ts") {
+            var dot = path.lastIndexOf(".", path.length - 4);
+            if (dot >= 0 && (path.substring(dot - 2, dot) === ".d" || path.substring(dot, dot + 2) === ".d")) {
+                return path;
+            }
+            return path.substring(0, path.length - 3) + ".js";
+        }
+        return path.replace(/(?<!\.d)\.[cm]ts$/, function (ext) { return ext === ".mts" ? ".mjs" : ".cjs"; });
+    }
+    return path;
+};
+// Rewrite
+import {} from "./foo.js";
+import {} from "../foo.mjs";
+import {} from "../../foo.cjs";
+import {} from "./foo.jsx";
+const foo = require("./foo.js");
+import "./foo.js";
+export * from "./foo.js";
+//Shim
+import(__rewriteRelativeImportExtension("./foo.ts", true));
+import(__rewriteRelativeImportExtension("./foo.ts", true), { with: { attr: "value" } });
+import(__rewriteRelativeImportExtension("" + "./foo.ts", true));
+//// [js.js]
+var __rewriteRelativeImportExtension = (this && this.__rewriteRelativeImportExtension) || function (path, preserveJsx) {
+    if (typeof path === "string" && path[0] === "." && (path[1] === "/" || path[1] === "." && path[2] === "/")) {
+        if (path.substring(path.length - 4) === ".tsx") {
+            return path.substring(0, path.length - 4) + (preserveJsx ? ".jsx" : ".js");
+        }
+        if (path.substring(path.length - 3) === ".ts") {
+            var dot = path.lastIndexOf(".", path.length - 4);
+            if (dot >= 0 && (path.substring(dot - 2, dot) === ".d" || path.substring(dot, dot + 2) === ".d")) {
+                return path;
+            }
+            return path.substring(0, path.length - 3) + ".js";
+        }
+        return path.replace(/(?<!\.d)\.[cm]ts$/, function (ext) { return ext === ".mts" ? ".mjs" : ".cjs"; });
+    }
+    return path;
+};
+// Rewrite
+import {} from "./foo.js";
+import {} from "../foo.mjs";
+import {} from "../../foo.cjs";
+import {} from "./foo.jsx";
+import "./foo.js";
+export * from "./foo.js";
+// Shim
+import(__rewriteRelativeImportExtension("./foo.ts", true));

```

Comment:
Minor potential UX optimization: You think it'd be worth building in knowledge in the checker/binder that `import(__rewriteRelativeImportExtension("./foo.ts"))` can be bound/checked the same as `import("./foo.ts")`, just so LS operations on emitted files work a bit better?

Reply:
Yeah, I guess I thought we'd not emit the helper when we know statically that it's a literal, right?

Reply:
> Yeah, I guess I thought we'd not emit the helper when we know statically that it's a literal, right?

It was suggested that import/require calls be shimmed unconditionally so someone debugging the emitted code can understand what happened

Reply:
But, _not_ taking that suggestion would be a much easier way of making the emitted code work in the LS. I’m a bit concerned about how far the increased complexity of changing what imports/requires can be considered a module reference would reach into the LS.

Reply:
🤷 Maybe one day we'll change `require`/`import` inference logic to operate on resolved types and not just literals, so declaring `declare function __rewriteRelativeImportExtension<T extends string>(arg: T): T;` makes an `import`/`resolve` of the call on the literal still work (since the function is defined as a pass-thru, type-wise) and we shouldn't bother special-casing it now.

Reply:
I’m going to push a commit that changes the literal case from shimming to rewriting—I think losing the explicitness of what happened in the emit is the least bad of the tradeoffs discussed here. We can always emit a synthesized comment if we want to make it more clear that something happened in the emit. (Might not be a bad idea for all the rewrites.) I won’t have time to get to that before I leave, but can come back to it for the RC if that’s something we want to do.